File: DiscordConnector/DiscordDatabaseController/controller.py
"""DiscordDatabaseController implementation."""

from sqlalchemy import select
from sqlalchemy.orm import joinedload

from DiscordConnector.DiscordDatabaseController.database import Database
from DiscordConnector.DiscordDatabaseController.interface import (
    IDiscordDatabaseController,
    User,
    Role,
    Category,
    Channel,
    DatabaseError,
)
from DiscordConnector.DiscordDatabaseController.models import (
    CategoryModel,
    ChannelModel,
    RoleModel,
    UserModel,
)


class DiscordDatabaseController(IDiscordDatabaseController):
    """Database controller for Discord server state management."""

    def __init__(self, database_url: str):
        """Initialize controller with database URL."""
        self._db = Database(database_url)

    async def connect(self) -> None:
        """Initialize database connection and create tables."""
        await self._db.connect()

    async def disconnect(self) -> None:
        """Close database connection."""
        await self._db.disconnect()

    async def __aenter__(self) -> "DiscordDatabaseController":
        await self.connect()
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.disconnect()

    # Helper methods for conversion
    def _model_to_user(self, model: UserModel) -> User:
        return User(
            discord_user_id=model.discord_user_id,
            display_name=model.display_name,
            member_id=model.member_id,
            role_ids=[r.role_id for r in model.roles] if model.roles else None,
        )

    def _model_to_role(self, model: RoleModel) -> Role:
        return Role(
            role_id=model.role_id,
            role_name=model.role_name,
            permissions=0,
        )

    def _model_to_channel(self, model: ChannelModel) -> Channel:
        return Channel(
            channel_id=model.channel_id,
            channel_name=model.channel_name,
            category_id=model.category_id,
            role_ids=[r.role_id for r in model.roles] if model.roles else None,
        )

    def _model_to_category(self, model: CategoryModel) -> Category:
        return Category(
            category_id=model.category_id,
            category_name=model.category_name,
            channels=[self._model_to_channel(c) for c in model.channels] if model.channels else None,
            role_ids=[r.role_id for r in model.roles] if model.roles else None,
        )

    # User CRUD
    async def get_users(self, member_id: str | None = None) -> list[User]:
        async with self._db.session() as session:
            query = select(UserModel).options(joinedload(UserModel.roles))
            if member_id:
                query = query.filter(UserModel.member_id == member_id)
            result = await session.execute(query)
            users = result.unique().scalars().all()
            return [self._model_to_user(u) for u in users]

    async def get_user(self, discord_user_id: str) -> User | None:
        async with self._db.session() as session:
            result = await session.execute(
                select(UserModel)
                .options(joinedload(UserModel.roles))
                .filter(UserModel.discord_user_id == discord_user_id)
            )
            user = result.unique().scalar_one_or_none()
            return self._model_to_user(user) if user else None

    async def create_user(
        self,
        discord_user_id: str,
        display_name: str,
        member_id: str | None = None,
    ) -> User:
        async with self._db.session() as session:
            # Check if user already exists
            existing = await session.execute(
                select(UserModel).filter(UserModel.discord_user_id == discord_user_id)
            )
            if existing.scalar_one_or_none():
                raise DatabaseError(f"User {discord_user_id} already exists")

            user = UserModel(
                discord_user_id=discord_user_id,
                display_name=display_name,
                member_id=member_id,
            )
            session.add(user)
            await session.flush()
            # Return without accessing roles (new user has no roles)
            return User(
                discord_user_id=user.discord_user_id,
                display_name=user.display_name,
                member_id=user.member_id,
                role_ids=[],
            )

    async def update_user(
        self,
        discord_user_id: str,
        display_name: str,
        member_id: str | None = None,
    ) -> User | None:
        async with self._db.session() as session:
            result = await session.execute(
                select(UserModel)
                .options(joinedload(UserModel.roles))
                .filter(UserModel.discord_user_id == discord_user_id)
            )
            user = result.unique().scalar_one_or_none()
            if not user:
                return None

            user.display_name = display_name
            user.member_id = member_id
            await session.flush()
            return self._model_to_user(user)

    async def delete_user(self, discord_user_id: str) -> bool:
        async with self._db.session() as session:
            result = await session.execute(
                select(UserModel).filter(UserModel.discord_user_id == discord_user_id)
            )
            user = result.scalar_one_or_none()
            if not user:
                return False

            await session.delete(user)
            return True

    async def sync_user_roles(
        self,
        discord_user_id: str,
        role_ids: list[str],
    ) -> int:
        async with self._db.session() as session:
            result = await session.execute(
                select(UserModel)
                .options(joinedload(UserModel.roles))
                .filter(UserModel.discord_user_id == discord_user_id)
            )
            user = result.unique().scalar_one_or_none()
            if not user:
                raise DatabaseError(f"User {discord_user_id} not found")

            # Get roles
            roles_result = await session.execute(
                select(RoleModel).filter(RoleModel.role_id.in_(role_ids))
            )
            roles = roles_result.scalars().all()

            user.roles = list(roles)
            return len(roles)

    # Role CRUD
    async def get_roles(self) -> list[Role]:
        async with self._db.session() as session:
            result = await session.execute(select(RoleModel))
            roles = result.scalars().all()
            return [self._model_to_role(r) for r in roles]

    async def get_role(self, role_id: str) -> Role | None:
        async with self._db.session() as session:
            result = await session.execute(
                select(RoleModel).filter(RoleModel.role_id == role_id)
            )
            role = result.scalar_one_or_none()
            return self._model_to_role(role) if role else None

    async def create_role(
        self,
        role_id: str,
        role_name: str,
        permissions: int,
    ) -> Role:
        async with self._db.session() as session:
            role = RoleModel(
                role_id=role_id,
                role_name=role_name,
            )
            session.add(role)
            await session.flush()
            return Role(
                role_id=role.role_id,
                role_name=role.role_name,
                permissions=permissions,
            )

    async def update_role(
        self,
        role_id: str,
        role_name: str | None = None,
        permissions: int | None = None,
    ) -> Role | None:
        async with self._db.session() as session:
            result = await session.execute(
                select(RoleModel).filter(RoleModel.role_id == role_id)
            )
            role = result.scalar_one_or_none()
            if not role:
                return None

            if role_name is not None:
                role.role_name = role_name
            await session.flush()
            return Role(
                role_id=role.role_id,
                role_name=role.role_name,
                permissions=permissions if permissions is not None else 0,
            )

    async def delete_role(self, role_id: str) -> bool:
        async with self._db.session() as session:
            result = await session.execute(
                select(RoleModel).filter(RoleModel.role_id == role_id)
            )
            role = result.scalar_one_or_none()
            if not role:
                return False

            await session.delete(role)
            return True

    # Category CRUD
    async def get_categories(self) -> list[Category]:
        async with self._db.session() as session:
            result = await session.execute(
                select(CategoryModel).options(
                    joinedload(CategoryModel.channels).joinedload(ChannelModel.roles),
                    joinedload(CategoryModel.roles),
                )
            )
            categories = result.unique().scalars().all()
            return [self._model_to_category(c) for c in categories]

    async def get_category(self, category_id: str) -> Category | None:
        async with self._db.session() as session:
            result = await session.execute(
                select(CategoryModel)
                .options(
                    joinedload(CategoryModel.channels).joinedload(ChannelModel.roles),
                    joinedload(CategoryModel.roles),
                )
                .filter(CategoryModel.category_id == category_id)
            )
            category = result.unique().scalar_one_or_none()
            return self._model_to_category(category) if category else None

    async def create_category(
        self,
        category_id: str,
        category_name: str,
    ) -> Category:
        async with self._db.session() as session:
            category = CategoryModel(
                category_id=category_id,
                category_name=category_name,
            )
            session.add(category)
            await session.flush()
            return Category(
                category_id=category.category_id,
                category_name=category.category_name,
                channels=[],
                role_ids=[],
            )

    async def delete_category(self, category_id: str) -> bool:
        async with self._db.session() as session:
            result = await session.execute(
                select(CategoryModel).filter(CategoryModel.category_id == category_id)
            )
            category = result.scalar_one_or_none()
            if not category:
                return False

            await session.delete(category)
            return True

    async def sync_category_permissions(
        self,
        category_id: str,
        role_ids: list[str],
    ) -> int:
        async with self._db.session() as session:
            result = await session.execute(
                select(CategoryModel)
                .options(joinedload(CategoryModel.roles))
                .filter(CategoryModel.category_id == category_id)
            )
            category = result.unique().scalar_one_or_none()
            if not category:
                raise DatabaseError(f"Category {category_id} not found")

            roles_result = await session.execute(
                select(RoleModel).filter(RoleModel.role_id.in_(role_ids))
            )
            roles = roles_result.scalars().all()

            category.roles = list(roles)
            return len(roles)

    # Channel CRUD
    async def get_channels(self) -> list[Channel]:
        async with self._db.session() as session:
            result = await session.execute(
                select(ChannelModel).options(joinedload(ChannelModel.roles))
            )
            channels = result.unique().scalars().all()
            return [self._model_to_channel(c) for c in channels]

    async def get_channel(self, channel_id: str) -> Channel | None:
        async with self._db.session() as session:
            result = await session.execute(
                select(ChannelModel)
                .options(joinedload(ChannelModel.roles))
                .filter(ChannelModel.channel_id == channel_id)
            )
            channel = result.unique().scalar_one_or_none()
            return self._model_to_channel(channel) if channel else None

    async def create_channel(
        self,
        channel_id: str,
        channel_name: str,
        category_id: str,
        allowed_role_ids: list[str] | None = None,
    ) -> Channel:
        async with self._db.session() as session:
            # Verify category exists
            cat_result = await session.execute(
                select(CategoryModel).filter(CategoryModel.category_id == category_id)
            )
            if not cat_result.scalar_one_or_none():
                raise DatabaseError(f"Category {category_id} not found")

            channel = ChannelModel(
                channel_id=channel_id,
                channel_name=channel_name,
                category_id=category_id,
            )

            role_ids_result: list[str] = []
            if allowed_role_ids:
                roles_result = await session.execute(
                    select(RoleModel).filter(RoleModel.role_id.in_(allowed_role_ids))
                )
                roles = list(roles_result.scalars().all())
                channel.roles = roles
                role_ids_result = [r.role_id for r in roles]

            session.add(channel)
            await session.flush()
            return Channel(
                channel_id=channel.channel_id,
                channel_name=channel.channel_name,
                category_id=channel.category_id,
                role_ids=role_ids_result,
            )

    async def delete_channel(self, channel_id: str) -> bool:
        async with self._db.session() as session:
            result = await session.execute(
                select(ChannelModel).filter(ChannelModel.channel_id == channel_id)
            )
            channel = result.scalar_one_or_none()
            if not channel:
                return False

            await session.delete(channel)
            return True

    async def sync_channel_permissions(
        self,
        channel_id: str,
        role_ids: list[str],
    ) -> int:
        async with self._db.session() as session:
            result = await session.execute(
                select(ChannelModel)
                .options(joinedload(ChannelModel.roles))
                .filter(ChannelModel.channel_id == channel_id)
            )
            channel = result.unique().scalar_one_or_none()
            if not channel:
                raise DatabaseError(f"Channel {channel_id} not found")

            roles_result = await session.execute(
                select(RoleModel).filter(RoleModel.role_id.in_(role_ids))
            )
            roles = roles_result.scalars().all()

            channel.roles = list(roles)
            return len(roles)

File: DiscordConnector/src/DiscordDatabaseController/controller.ts
import pg from "pg";
import { DatabaseError } from "../errors";
import type { DbCategory, DbChannel, DbRole, DbUser, Snowflake } from "../types";
import type { IDiscordDatabaseController } from "./interface";

const { Client } = pg;

type Row = Record<string, unknown>;

export class DiscordDatabaseController implements IDiscordDatabaseController {
  constructor(private readonly env: Env) {}

  async connect(): Promise<void> {}

  async disconnect(): Promise<void> {}

  async getUsers(memberId?: string | null): Promise<DbUser[]> {
    const rows = await this.query<Row>(
      `select u.discord_user_id, u.display_name, u.member_id,
              coalesce(array_agg(ur.role_id) filter (where ur.role_id is not null), '{}') as role_ids
         from users u
         left join user_role ur on ur.discord_user_id = u.discord_user_id
        where ($1::text is null or u.member_id::text = $1)
        group by u.discord_user_id, u.display_name, u.member_id
        order by u.discord_user_id`,
      [memberId ?? null],
    );
    return rows.map(userFromRow);
  }

  async getUser(discordUserId: Snowflake): Promise<DbUser | null> {
    const users = await this.getUsersByIds([discordUserId]);
    return users[0] ?? null;
  }

  async createUser(discordUserId: Snowflake, displayName: string, memberId: string | null = null): Promise<DbUser> {
    try {
      await this.query(
        "insert into users (discord_user_id, display_name, member_id) values ($1, $2, $3)",
        [discordUserId, displayName, memberId],
      );
    } catch (error) {
      if (isPgUniqueViolation(error)) {
        throw new DatabaseError(`User ${discordUserId} already exists`);
      }
      throw error;
    }
    return { discordUserId, displayName, memberId, roleIds: [] };
  }

  async updateUser(discordUserId: Snowflake, displayName: string, memberId: string | null = null): Promise<DbUser | null> {
    const rows = await this.query<Row>(
      "update users set display_name = $2, member_id = $3 where discord_user_id = $1 returning discord_user_id, display_name, member_id",
      [discordUserId, displayName, memberId],
    );
    if (rows.length === 0) {
      return null;
    }
    return this.getUser(discordUserId);
  }

  async deleteUser(discordUserId: Snowflake): Promise<boolean> {
    const rows = await this.query<Row>("delete from users where discord_user_id = $1 returning discord_user_id", [discordUserId]);
    return rows.length > 0;
  }

  async syncUserRoles(discordUserId: Snowflake, roleIds: Snowflake[]): Promise<number> {
    if (await this.getUser(discordUserId) === null) {
      throw new DatabaseError(`User ${discordUserId} not found`);
    }
    await this.transaction(async (client) => {
      await client.query("delete from user_role where discord_user_id = $1", [discordUserId]);
      for (const roleId of roleIds) {
        await client.query(
          `insert into user_role (discord_user_id, role_id)
           select $1, role_id from roles where role_id = $2
           on conflict do nothing`,
          [discordUserId, roleId],
        );
      }
    });
    return this.countUserRoles(discordUserId);
  }

  async getRoles(): Promise<DbRole[]> {
    const rows = await this.query<Row>("select role_id, role_name from roles order by role_id");
    return rows.map(roleFromRow);
  }

  async getRole(roleId: Snowflake): Promise<DbRole | null> {
    const rows = await this.query<Row>("select role_id, role_name from roles where role_id = $1", [roleId]);
    return rows[0] === undefined ? null : roleFromRow(rows[0]);
  }

  async createRole(roleId: Snowflake, roleName: string, permissions: number): Promise<DbRole> {
    await this.query("insert into roles (role_id, role_name) values ($1, $2)", [roleId, roleName]);
    return { roleId, roleName, permissions };
  }

  async updateRole(roleId: Snowflake, roleName: string | null = null, permissions: number | null = null): Promise<DbRole | null> {
    const existing = await this.getRole(roleId);
    if (existing === null) {
      return null;
    }
    const nextName = roleName ?? existing.roleName;
    await this.query("update roles set role_name = $2 where role_id = $1", [roleId, nextName]);
    return { roleId, roleName: nextName, permissions: permissions ?? 0 };
  }

  async deleteRole(roleId: Snowflake): Promise<boolean> {
    const rows = await this.query<Row>("delete from roles where role_id = $1 returning role_id", [roleId]);
    return rows.length > 0;
  }

  async getCategories(): Promise<DbCategory[]> {
    const rows = await this.query<Row>("select category_id, category_name from categories order by category_id");
    const categories = await Promise.all(rows.map(categoryFromRow).map(async (category) => ({
      ...category,
      channels: await this.getChannelsByCategory(category.categoryId),
      roleIds: await this.getRoleIds("category_role", "category_id", category.categoryId),
    })));
    return categories;
  }

  async getCategory(categoryId: Snowflake): Promise<DbCategory | null> {
    const rows = await this.query<Row>("select category_id, category_name from categories where category_id = $1", [categoryId]);
    if (rows.length === 0) {
      return null;
    }
    const category = categoryFromRow(rows[0]!);
    return {
      ...category,
      channels: await this.getChannelsByCategory(categoryId),
      roleIds: await this.getRoleIds("category_role", "category_id", categoryId),
    };
  }

  async createCategory(categoryId: Snowflake, categoryName: string): Promise<DbCategory> {
    await this.query("insert into categories (category_id, category_name) values ($1, $2)", [categoryId, categoryName]);
    return { categoryId, categoryName, channels: [], roleIds: [] };
  }

  async deleteCategory(categoryId: Snowflake): Promise<boolean> {
    const rows = await this.query<Row>("delete from categories where category_id = $1 returning category_id", [categoryId]);
    return rows.length > 0;
  }

  async syncCategoryPermissions(categoryId: Snowflake, roleIds: Snowflake[]): Promise<number> {
    if (await this.getCategory(categoryId) === null) {
      throw new DatabaseError(`Category ${categoryId} not found`);
    }
    await this.syncRoles("category_role", "category_id", categoryId, roleIds);
    return this.getRoleIds("category_role", "category_id", categoryId).then((ids) => ids.length);
  }

  async getChannels(): Promise<DbChannel[]> {
    const rows = await this.query<Row>("select channel_id, channel_name, category_id from channels order by channel_id");
    return Promise.all(rows.map((row) => this.channelFromRowWithRoles(row)));
  }

  async getChannel(channelId: Snowflake): Promise<DbChannel | null> {
    const rows = await this.query<Row>("select channel_id, channel_name, category_id from channels where channel_id = $1", [channelId]);
    return rows[0] === undefined ? null : this.channelFromRowWithRoles(rows[0]);
  }

  async createChannel(channelId: Snowflake, channelName: string, categoryId: Snowflake, allowedRoleIds: Snowflake[] | null = null): Promise<DbChannel> {
    if (await this.getCategory(categoryId) === null) {
      throw new DatabaseError(`Category ${categoryId} not found`);
    }
    await this.transaction(async (client) => {
      await client.query(
        "insert into channels (channel_id, channel_name, category_id) values ($1, $2, $3)",
        [channelId, channelName, categoryId],
      );
      for (const roleId of allowedRoleIds ?? []) {
        await client.query(
          `insert into channel_role (channel_id, role_id)
           select $1, role_id from roles where role_id = $2
           on conflict do nothing`,
          [channelId, roleId],
        );
      }
    });
    return { channelId, channelName, categoryId, roleIds: await this.getRoleIds("channel_role", "channel_id", channelId) };
  }

  async deleteChannel(channelId: Snowflake): Promise<boolean> {
    const rows = await this.query<Row>("delete from channels where channel_id = $1 returning channel_id", [channelId]);
    return rows.length > 0;
  }

  async syncChannelPermissions(channelId: Snowflake, roleIds: Snowflake[]): Promise<number> {
    if (await this.getChannel(channelId) === null) {
      throw new DatabaseError(`Channel ${channelId} not found`);
    }
    await this.syncRoles("channel_role", "channel_id", channelId, roleIds);
    return this.getRoleIds("channel_role", "channel_id", channelId).then((ids) => ids.length);
  }

  private async getUsersByIds(ids: Snowflake[]): Promise<DbUser[]> {
    if (ids.length === 0) {
      return [];
    }
    const rows = await this.query<Row>(
      `select u.discord_user_id, u.display_name, u.member_id,
              coalesce(array_agg(ur.role_id) filter (where ur.role_id is not null), '{}') as role_ids
         from users u
         left join user_role ur on ur.discord_user_id = u.discord_user_id
        where u.discord_user_id = any($1::text[])
        group by u.discord_user_id, u.display_name, u.member_id`,
      [ids],
    );
    return rows.map(userFromRow);
  }

  private async countUserRoles(discordUserId: Snowflake): Promise<number> {
    const rows = await this.query<Row>("select count(*)::int as count from user_role where discord_user_id = $1", [discordUserId]);
    return Number(rows[0]?.count ?? 0);
  }

  private async getChannelsByCategory(categoryId: Snowflake): Promise<DbChannel[]> {
    const rows = await this.query<Row>("select channel_id, channel_name, category_id from channels where category_id = $1 order by channel_id", [categoryId]);
    return Promise.all(rows.map((row) => this.channelFromRowWithRoles(row)));
  }

  private async channelFromRowWithRoles(row: Row): Promise<DbChannel> {
    const channel = channelFromRow(row);
    return { ...channel, roleIds: await this.getRoleIds("channel_role", "channel_id", channel.channelId) };
  }

  private async getRoleIds(table: "user_role" | "category_role" | "channel_role", idColumn: string, id: Snowflake): Promise<Snowflake[]> {
    const rows = await this.query<Row>(`select role_id from ${table} where ${idColumn} = $1 order by role_id`, [id]);
    return rows.map((row) => String(row.role_id));
  }

  private async syncRoles(table: "category_role" | "channel_role", idColumn: string, id: Snowflake, roleIds: Snowflake[]): Promise<void> {
    await this.transaction(async (client) => {
      await client.query(`delete from ${table} where ${idColumn} = $1`, [id]);
      for (const roleId of roleIds) {
        await client.query(
          `insert into ${table} (${idColumn}, role_id)
           select $1, role_id from roles where role_id = $2
           on conflict do nothing`,
          [id, roleId],
        );
      }
    });
  }

  private async query<T extends Row>(sql: string, values: unknown[] = []): Promise<T[]> {
    const client = new Client({ connectionString: this.env.HYPERDRIVE.connectionString });
    await client.connect();
    try {
      const result = await client.query<T>(sql, values);
      return result.rows;
    } finally {
      await client.end();
    }
  }

  private async transaction(action: (client: pg.Client) => Promise<void>): Promise<void> {
    const client = new Client({ connectionString: this.env.HYPERDRIVE.connectionString });
    await client.connect();
    try {
      await client.query("begin");
      await action(client);
      await client.query("commit");
    } catch (error) {
      await client.query("rollback");
      throw error;
    } finally {
      await client.end();
    }
  }
}

function userFromRow(row: Row): DbUser {
  return {
    discordUserId: String(row.discord_user_id),
    displayName: String(row.display_name),
    memberId: row.member_id === null ? null : String(row.member_id),
    roleIds: Array.isArray(row.role_ids) ? row.role_ids.map(String) : [],
  };
}

function roleFromRow(row: Row): DbRole {
  return {
    roleId: String(row.role_id),
    roleName: String(row.role_name),
    permissions: 0,
  };
}

function categoryFromRow(row: Row): DbCategory {
  return {
    categoryId: String(row.category_id),
    categoryName: String(row.category_name),
    channels: [],
    roleIds: [],
  };
}

function channelFromRow(row: Row): DbChannel {
  return {
    channelId: String(row.channel_id),
    channelName: String(row.channel_name),
    categoryId: String(row.category_id),
    roleIds: [],
  };
}

function isPgUniqueViolation(error: unknown): boolean {
  return typeof error === "object" && error !== null && "code" in error && error.code === "23505";
}
